message/messaging_service.cc

```diff
@@ -606,18 +606,26 @@ gms::inet_address messaging_service::get_preferred_ip(gms::inet_address ep) {
 
 void messaging_service::init_local_preferred_ip_cache(const std::unordered_map<gms::inet_address, gms::inet_address>& ips_cache) {
     _preferred_ip_cache = ips_cache;
+    _preferred_to_endpoint.clear();
     //
     // Reset the connections to the endpoints that have entries in
     // _preferred_ip_cache so that they reopen with the preferred IPs we've
     // just read.
     //
     for (auto& p : _preferred_ip_cache) {
         this->remove_rpc_client(msg_addr(p.first));
+        _preferred_to_endpoint[p.second] = p.first;
```

Comment:
There's an existing problem here that the cache isn't maintained. So if a node is removed and then a new node is added with one of the addresses matching but the other not, we get incorrect mappings. We should look into invalidating the cache when a node is removed.

Reply:
messaging_service::init_local_preferred_ip_cache is called only once during startup. The code in messaging_service::init_local_preferred_ip_cache looks correct.  How can we end up that a node is only in one of the mappings ( _preferred_ip_cache and _preferred_to_endpoint)? The mapping is updated only in messaging_service::cache_preferred_ip() where both of the mappings are updated at the same time.

Reply:
@avikivity - as @asias says, there is no way to have the cache contain mismatched public/private ip pairs. The only thing possible would be for a node 1 with public ip A and private ip B to go down, a new node 2 go up using public ip C but private ip B, in which case, until the cache is updated we _could_ accept an unencrypted call from B and assume it is node 1. 
However: 
1.) It is still a node on the private network. Given that the actual usage here is more or less limited to ec2 multi and `gossiping_property_file_snitch` (only ones allowing this) it seems to be reasonably secure - at worst we allow non-TLS a bit
2.) This should never happen, since the first messages from gossip are _always_ TLS if any TLS in enabled, and the first things that should happen is node join -> update the cache. 

It would have to be a malignant actor on the private network faking gossip (perfectly up till now), gaining an unencrypted connection. Which seems pointless, since in that case he already has access to all the data he reads...

Reply:
A new node starts, we call cache_preferred_ip() to add an entry to the cache.

The node is removed, we don't remove it from the cache.

A different node with the same address is started, it is already in the cache.

Reply:
> A new node starts, we call cache_preferred_ip() to add an entry to the cache.
> 
> The node is removed, we don't remove it from the cache.
> 
> A different node with the same address is started, it is already in the cache.

We have

```
 621 void messaging_service::cache_preferred_ip(gms::inet_address ep, gms::inet_address ip) {
 622     _preferred_ip_cache[ep] = ip;
 623     _preferred_to_endpoint[ip] = ep;
 624 }
```

which sets both the  public to private and private to public mapping at the same time.

Are you talking about

Old mapping
8.8.8.8 -> 192.168.1.8
192.168.1.8-> 8.8.8.8

New mapping
8.8.8.8 -> 192.168.1.9
192.168.1.9-> 8.8.8.8
192.168.1.8-> 8.8.8.8

The old entry 192.168.1.8-> 8.8.8.8 is still there. 

But is this a problem? We won't use the old address 192.168.1.8 anyway.

Reply:
Perhaps we can construct the private to public mapping from scratch all the time when messaging_service::cache_preferred_ip is called if we think the public to private mapping is always correct. 

Anyway, Avi, if you think this is an existing problem, let's not block this PR which blocks 4.6.

Reply:
I'm not at all sure about what's going on. Anyway, we should not use a cache. Caches are great when the amount of data is too large to hold in memory at the same time. Here we should just have a map, and continuously maintain it (i.e. it is a bug to look up in the map and not find an entry).